### PR TITLE
Semantics of x86 uplifting: reduce 'Repeat' ops

### PR DESCRIPTION
**Pull request by kui8shi**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
The current implementation of uplifting x86 assemblies heavily relies on the 'Repeat' operation which is hard to handle statically. So I modified the lifting code to reduce the use of  'Repeat' ops. The main target is the parity flag calculation, which appears in most arithmetic x86 ops. It works well in normal execution, however in the static analysis view, we can't (simply) know how many times the loop is repeated, hence it is required to divide a single x86 op into multiple basic blocks.

Key Changes:
- Replaced Repeat with bitwise operations for a parity calculation.
- Simplified rotate and shift calculations.
- Updated affected test cases.

...

**Test plan**
rz-tracetest:
```sh
File: result_after_commit.txt
--------------------------------------
              success: 112486  97.94%
              skipped: 0        0.00%
             unlifted: 1015     0.88%
           invalid il: 0        0.00%
     vm runtime error: 0        0.00%
          misexecuted: 1353     1.18%


File: result_before_commit.txt
--------------------------------------
              success: 112368  97.84%
              skipped: 0        0.00%
             unlifted: 1015     0.88%
           invalid il: 0        0.00%
     vm runtime error: 0        0.00%
          misexecuted: 1471     1.28%
```

~TODO: Fix all affected tests~

...

**Closing issues**

...

### CODE REVIEW

**Review by DMaroo — COMMENTED** [review]
This seems fine by me, thanks! Will approve once the CI is green.

File: librz/arch/isa/x86/common.c [hunk 1]
```diff
@@ -889,35 +889,21 @@ RZ_IPI RzILOpBool *x86_il_is_sub_underflow(RZ_OWN RZ_NONNULL RzILOpPure *res, RZ
...
-	};
 
+	// A parity calculation can be reduced to nested shift-right and xor operations.
+	// see https://github.com/BinaryAnalysisPlatform/bap/blob/master/plugins/x86/x86_lifter.ml#L215
+	RzILOpPure *accumulate = LET("_val", val,
+		LET("_c4", LOGXOR(VARLP("_val"), SHIFTR0(VARLP("_val"), U8(4))),
+			LET("_c2", LOGXOR(VARLP("_c4"), SHIFTR0(VARLP("_c4"), U8(2))),
+				LOGXOR(VARLP("_c2"), SHIFTR0(VARLP("_c2"), U8(1))))));
+	RzILOpPure *ret = INV(LSB(accumulate));
 	return ret;
```
Comment (wargio) [inline librz/arch/isa/x86/common.c]: just return

File: librz/arch/isa/x86/il_ops.inc [hunk 2]
```diff
@@ -1740,30 +1741,20 @@ IL_LIFTER(pushal) {
...
-		cnt_masked = SETL("_cnt_mask", UNSIGNED(5, x86_il_get_op_implicit(1, size))); \
-		tmp_count_size = 5; \
+		shift = SETL("_shift", MOD(UNSIGNED(shift_size, x86_il_get_op_implicit(1, size_in_bytes)), UN(shift_size, shift_max))); \
 		break; \
 	case 4: \
-		temp_count = SETL("_tmp_cnt", UNSIGNED(5, x86_il_get_op_implicit(1, size))); \
-		cnt_masked = SETL("_cnt_mask", UNSIGNED(5, x86_il_get_op_implicit(1, size))); \
-		tmp_count_size = 5; \
-		break; \
 	case 8: \
```
Comment (wargio) [inline librz/arch/isa/x86/il_ops.inc]: if case 4 and case 8 are the same, then add a `/* fall-thru */` comment..

File: librz/arch/isa/x86/il_ops.inc [hunk 3]
```diff
@@ -1740,30 +1741,20 @@ IL_LIFTER(pushal) {
...
+	ut8 shift_max = size + 1; \
+	RzILOpEffect *pre_dest = SETL("_pre_dest", x86_il_get_op(0)); \
+	RzILOpEffect *shift = NULL; \
+	switch (size_in_bytes) { \
 	case 1: \
-		temp_count = SETL("_tmp_cnt", MOD(UNSIGNED(5, x86_il_get_op_implicit(1, size)), UN(5, 9))); \
-		cnt_masked = SETL("_cnt_mask", UNSIGNED(5, x86_il_get_op_implicit(1, size))); \
-		tmp_count_size = 5; \
-		break; \
 	case 2: \
```
Comment (wargio) [inline librz/arch/isa/x86/il_ops.inc]: if they are the same, add `/* fall-thru */` comment.

**Review by DMaroo — APPROVED** [review]
Great work! I like the changes. The code looks much better now.

**Comment by XVilka** [comment]
Thanks for the PR!

**Pull request merged**